### PR TITLE
Refactor testerina core ballerina module

### LINKED ISSUE #42123: [Improvement]: Improve the quality of Ballerina code in testerina ballerina code
### Description

There is major room for improvement in the testerina ballerina code quality when it comes to best coding practices. We need to go through the code base and improve the code as much as possible.

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Test Framework

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gayaldassanayake**
## Purpose
$ subject

Fixes #42123 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/testerina/modules/testerina-core/src/main/ballerina/annotation_processor.bal [hunk 1]
```diff
@@ -72,31 +79,39 @@ function processConfigAnnotation(string name, function f) returns boolean {
...
+                && (filterDisableGroups.length() == 0 || !hasGroup(config.groups, filterDisableGroups))
+                && hasTest(name);
         config.groups.forEach('group => groupStatusRegistry.incrementTotalTest('group, enabled));
         dataDrivenTestParams[name] = params;
+
+        // if the serial execution field is set, or if not parallelizable or if parallel execution is disabled, then
+        // set the serial execution to true.
+        boolean serialExecution = config?.serialExecution != ()
+                || !isSatisfiedParallelizableConditions
+                || !executionManager.isParallelExecutionEnabled();
```
Comment (SasinduDilshara) [inline misc/testerina/modules/testerina-core/src/main/ballerina/annotation_processor.bal:110]: imo, this needs to be come as the second condion and L109 should be the 3rd

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal [hunk 2]
```diff
@@ -238,34 +238,35 @@ public isolated function assertFail(string msg = "Test Failed!") returns never {
...
+        string diff = getMapValueDiff(<map<anydata>>actual, <map<anydata>>expected);
+        errorMsg = string `${msg}` + "\n \nexpected: " + string `'${expectedStr}'` + "\nactual\t: " +
                             string `'${actualStr}'` + "\n \nDiff\t:\n" + string `${diff}`;
-        } else {
-            errorMsg = string `${msg}` + "\n \nexpected: " + string `'${expectedStr}'` + "\nactual\t: "
-                                                 + string `'${actualStr}'`;
-        }
-        return errorMsg;
+    } else {
+        errorMsg = string `${msg}` + "\n \nexpected: " + string `'${expectedStr}'` + "\nactual\t: "
```
Comment (SasinduDilshara) [inline misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal:266]: Can't we use a single string template

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal [hunk 3]
```diff
@@ -309,48 +310,48 @@ isolated function getMapValueDiff(map<anydata> actualMap, map<anydata> expectedM
...
-                    string childDiff;
-                    if (expectedChildVal != actualChildVal) {
-                        [childDiff, diffCount] = getValueComparison(actualChildVal, expectedChildVal, keyVal + "." + childKey, diffCount);
-                        if (diffCount != (orderCount + 1)) {
-                            diff = diff.concat("\n");
-                        }
-                        diff = diff.concat(childDiff);
+    } else if actual is map<anydata> && expected is map<anydata> {
+        string[] expectedkeyArray = (<map<anydata>>expected).keys();
+        string[] actualKeyArray = (<map<anydata>>actual).keys();
```
Comment (LakshanWeerasinghe) [inline misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal:333]:
```suggestion
        string[] expectedkeyArray = expected.keys();
        string[] actualKeyArray = actual.keys();
```
The type narrowing should work here.

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal (hunk 2 above)
Comment (gayaldassanayake) [reply misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal]: Fixed with 27c521a5975fb3e867c8bdc46b9cc676e14f992c

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal [hunk 4]
```diff
@@ -309,48 +310,48 @@ isolated function getMapValueDiff(map<anydata> actualMap, map<anydata> expectedM
...
             }
-        } else {
-            diff = diff.concat("\n", "key: ", keyVal, "\n \nexpected value\t: ", expected.toString(),
-            "\nactual value\t: ", actual.toString());
-            diffCount = diffCount + 1;
+
         }
+    } else {
+        diff = diff.concat("\n", "key: ", keyVal, "\n \nexpected value\t: ", expected.toString(),
+        "\nactual value\t: ", actual.toString());
```
[the remark this answers is not in the document]
Comment (gayaldassanayake) [reply misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal]: Fixed with 27c521a5975fb3e867c8bdc46b9cc676e14f992c

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal (hunk 3 above)
Comment (gayaldassanayake) [reply misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal]: Changed with 27c521a5975fb3e867c8bdc46b9cc676e14f992c

File: misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal [hunk 5]
```diff
@@ -309,48 +310,48 @@ isolated function getMapValueDiff(map<anydata> actualMap, map<anydata> expectedM
...
-                        [childDiff, diffCount] = getValueComparison(actualChildVal, expectedChildVal, keyVal + "." + childKey, diffCount);
-                        if (diffCount != (orderCount + 1)) {
-                            diff = diff.concat("\n");
-                        }
-                        diff = diff.concat(childDiff);
+    } else if actual is map<anydata> && expected is map<anydata> {
+        string[] expectedkeyArray = (<map<anydata>>expected).keys();
+        string[] actualKeyArray = (<map<anydata>>actual).keys();
+        int orderCount = diffCount;
+        foreach string childKey in actualKeyArray {
```
[the remark this answers is not in the document]
Comment (gayaldassanayake) [reply misc/testerina/modules/testerina-core/src/main/ballerina/assert.bal]: Fixed with 27c521a5975fb3e867c8bdc46b9cc676e14f992c

**Review by azinneera — APPROVED** [review]

**Pull request merged**